qemu/tests/queues_number_test.py

```diff
@@ -0,0 +1,211 @@
+            error_context.context("Enable multi queues in guest.", logging.info)
+            for nic in vm.virtnet:
+                ifname = utils_net.get_linux_ifname(session_serial, nic.mac)
+                queues = int(nic.queues)
+                change_queues_number(ifname, queues)
+            error_context.context("Run test %s background" % netperf_stress,
+                                  logging.info)
+            stress_thread = utils_misc.InterruptedThread(netperf_test)
+            stress_thread.start()
+            utils_misc.wait_for(lambda: wait_time, 0, 5,
+                                "Wait %s start background" % netperf_stress)
+
+            # ping test
+            error_context.context("Ping guest from host", logging.info)
+            args = (guest_ip, ping_time, ping_lost_ratio)
+            bg_ping = utils_misc.InterruptedThread(ping_test, args)
+            bg_ping.start()
+
+            error_context.context("Change queues number repeatedly",
+                                  logging.info)
+            repeat_counts = params.get_numeric("repeat_counts")
+            for nic in vm.virtnet:
+                queues = int(nic.queues)
+                if queues == 1:
+                    logging.info("Nic with single queue, skip and continue")
+                    continue
+                ifname = utils_net.get_linux_ifname(session_serial, nic.mac)
+                change_list = params.get("change_list").split(",")
+                for repeat_num in range(repeat_counts):
+                    error_context.context("Change queues number -- %sth"
+                                          % repeat_num, logging.info)
+                    queues_status = get_queues_status(ifname)
+                    for q_number in change_list:
+                        queues_status = change_queues_number(ifname,
+                                                             int(q_number),
+                                                             queues_status)
+
+            logging.info("wait for background test finish")
+            try:
+                stress_thread.join()
+            except Exception as err:
+                err_msg = "Run %s test background error!\n "
+                err_msg += "Error Info: '%s'"
+                test.error(err_msg % (netperf_stress, err))
+
+            logging.info("Wait for background ping test finish.")
+            try:
+                bg_ping.join()
+            except Exception as err:
+                txt = "Fail to wait background ping test finish. "
+                txt += "Got error message %s" % err
+                test.fail(txt)
+
+            if required_reboot:
+                logging.info("Rebooting guest ...")
+                vm.reboot()
+                required_reboot = False
+            else:
+                bg_test = False
+    finally:
+        n_server.package.env_cleanup(True)
+        n_client.package.env_cleanup(True)
```

Comment:
n_server.cleanup()
n_client.cleanup()

Reply:
Do we need invoke n_server.stop() and n_client.stop() first?

Reply:
No, This is a formal usage and the method can be found in the utils_netperf.py file.

Reply:
cleanup() will invoke this env_cleanup() function.
How about the stop() ? Do you need stop it at the end?

Reply:
package.env_cleanup() : "delete both netperf binary and source tarball", more in line with our requirements.
I have stop the netperf_server on L116. netperf_clinet will stop automatically after the operation is over.

Reply:
FYI:
```python
    def stop(self, target):
        if self.client == "nc":
            stop_cmd = "taskkill /F /IM %s*" % target
        else:
            stop_cmd = "killall %s" % target
        if self.is_target_running(target):
            self.session.cmd(stop_cmd, ignore_all_errors=True)
        if self.is_target_running(target):
            raise NetserverError("Cannot stop %s" % target)
        logging.info("Stop %s successfully" % target)

    def cleanup(self, clean_all=True):
        """
        Cleanup the netperf packages.

        :param clean_all: True to delete both netperf binary and source tarball
                          and False to delete only binary.
        """
        self.package.env_cleanup(clean_all=clean_all)
```